src/serve/hosting.ts

```diff
@@ -44,6 +45,34 @@ function startServer(options: any, config: any, port: number, init: TemplateServ
     stream: morganStream,
   });
 
+  const portInUse = () => {
+    const message = "Port " + options.port + " is not available.";
+    logger.log("WARN", clc.yellow("hosting: ") + message + " Trying another port...");
+    if (attempts < MAX_PORT_ATTEMPTS) {
+      // Another project that's running takes up to 4 ports: 1 hosting port and 3 functions ports
+      attempts++;
+      startServer(options, config, port + 5, init);
+    } else {
+      logger.log("WARN", message);
+      throw new FirebaseError("Could not find an open port for hosting development server.", {
+        exit: 1,
+      });
+    }
+  };
+
+  // On OSX, some ports may be reserved by the OS in a way that node http doesn't detect.
+  // Starting in MacOS 12.3 it does this with port 5000 our default port. This is a bad
+  // enough devexp that we should special case and ensure it's available.
+  if (process.platform === "darwin") {
+    try {
+      execSync(`lsof -i :${port}`);
```

Comment:
Looking through this, it seems like this only checks if the initial port is being used by osX, and will still break out if a subsequent attempt hits a osx occupied port. Should we include this `if (platform == 'darwin') lsof` in portInUse() as well?

Reply:
portInUse recurses, so it should work in subsequent attempts

Reply:
see line 54